solr/core/src/java/org/apache/solr/security/AllowListUrlChecker.java

```diff
@@ -0,0 +1,201 @@
+        } else if (System.getProperty("solr.disable.shardsWhitelist") != null) {
+            log.warn("Property 'solr.disable.shardsWhitelist' is deprecated, please use '{}' instead.", DISABLE_URL_ALLOW_LIST);
+        }
+        try {
+            return new AllowListUrlChecker(config.getAllowUrls());
+        } catch (MalformedURLException e) {
+            throw new SolrException(SolrException.ErrorCode.SERVER_ERROR,
+                    "Invalid URL syntax in '" + URL_ALLOW_LIST + "' configuration: " + config.getAllowUrls(), e);
+        }
+    }
+
+    /**
+     * @see #checkAllowList(List, ClusterState)
+     */
+    public void checkAllowList(List<String> urls) throws MalformedURLException {
+        checkAllowList(urls, null);
+    }
+
+    /**
+     * Checks that the given URLs are present in the configured allow-list or in the provided {@link ClusterState}
+     * (in case of cloud mode).
+     *
+     * @param urls         The list of urls to check.
+     * @param clusterState The up to date {@link ClusterState}, can be null in case of non-cloud mode.
+     * @throws MalformedURLException If an URL is invalid.
+     * @throws SolrException         If an URL is not present in the allow-list or in the provided {@link ClusterState}.
+     */
+    public void checkAllowList(List<String> urls, @Nullable ClusterState clusterState) throws MalformedURLException {
+        Set<String> clusterHostAllowList = clusterState == null ? Collections.emptySet() : clusterState.getHostAllowList();
+        for (String url : urls) {
+            String hostPort = parseHostPort(url);
+            if (!clusterHostAllowList.contains(hostPort) && !hostAllowList.contains(hostPort)) {
+                throw new SolrException(SolrException.ErrorCode.FORBIDDEN, "URL " + url +
+                        " is neither a live node of the cluster nor in the configured '" +
+                        URL_ALLOW_LIST + "' " + hostAllowList);
+            }
+        }
+    }
+
+    /**
+     * Whether this checker has been created with a non-empty allow-list of URLs.
+     */
+    public boolean hasExplicitAllowList() {
+        return !hostAllowList.isEmpty();
+    }
+
+    /**
+     * Whether the URL checking is enabled. Only {@link #ALLOW_ALL} returns false.
+     */
+    public boolean isEnabled() {
+        return true;
+    }
+
+    /**
+     * Only for testing.
+     */
+    @VisibleForTesting
+    public Set<String> getHostAllowList() {
+        return hostAllowList == null ? null : Collections.unmodifiableSet(hostAllowList);
+    }
+
+    @Override
+    public String toString() {
+        return getClass().getSimpleName() + " [allowList=" + hostAllowList + "]";
+    }
+
+    @VisibleForTesting
+    static Set<String> parseHostPorts(List<String> urls) throws MalformedURLException {
+        if (urls == null || urls.isEmpty()) {
+            return Collections.emptySet();
+        }
+        Set<String> hostPorts = new HashSet<>((int) (urls.size() / 0.7f) + 1);
```

Comment:
This feels like a weird optimization.

Reply:
I'll use Guava instead.

Reply:
I don't think that's necessary - just use the no-args HashSet constructor for now, unless we are convinced that this is a problem?

Reply:
Ok. I understand here the Set is small so we don't care that it enlarges and rehashes during the construction. I always try to initiallize with the right capacity when I have it ahead. It's sad that there is no such constructor in the jdk otherwise we would use it without questioning.

Reply:
The initial code you wrote `new HashSet<>((int) (urls.size() / 0.7f) + 1)` looked to me and Mike too complicated to warrant it; and I suspect it was incorrect, like you were over-thinking it based on your internal knowledge of hash sets/tables.  This constructor is merely the "initialCapacity".  If you want to use this constructor and you know how many items you want to add (you do), just pass `urls.size()`; no?
You might also use `Set.of(hostPorts.toArray())` which is a new JDK 10 feature and is leaner.  I learned about some of these the JDK APIs from this video https://www.youtube.com/watch?v=lwp2RZ__0ko and another presentation from Oracle.  Cool stuff!

Reply:
Unfortunately the HashSet(int) constructor capacity defines the capacity of the internal key array (internally a power of 2), which can only be loaded up to load factor (75%) before it enlarges and rehashes all keys. This means if we use new HashSet(urls.size()) we have high probability to rehash when adding the elements. I find this constructor misleading.
The formula I used is to ensure we don't reach the load factor. This is what is done in com.google.common.collect.Maps.capacity(int expectedSize), and what is used when we call Sets.newHashSetWithExpectedSize(int).

That said, here we talk about a small map so it is not even worth this discussion. That was my remark, I wished this newHashSetWithExpectedSize() was part of the jdk and we wouldn't talk about that.
Thanks for the link.

Reply:
I didn't know that!  Wow; that "initialCapacity" arg is trappy/misleading; I will not use it henceforth.  Using more of the new Set.of / Map.of will help avoid the discussion as well.